core/src/main/java/com/linecorp/armeria/common/stream/ByteStreamMessageOutputStream.java

```diff
@@ -0,0 +1,236 @@
+        this.outputStreamConsumer = outputStreamConsumer;
+        this.blockingTaskExecutor = blockingTaskExecutor;
+    }
+
+    @Override
+    public ByteStreamMessage range(long offset, long length) {
+        delegate.range(offset, length);
+        return this;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return delegate.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        return delegate.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return delegate.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        final Subscriber<HttpData> outputStreamSubscriber = new OutputStreamSubscriber(
+                subscriber, executor, outputStreamWriter, outputStreamConsumer, blockingTaskExecutor);
+        delegate.subscribe(outputStreamSubscriber, executor, options);
+    }
+
+    @Override
+    public void abort() {
+        delegate.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        delegate.abort(cause);
+    }
+
+    private static final class OutputStreamSubscriber implements Subscriber<HttpData>, Subscription {
+
+        private final Subscriber<? super HttpData> downstream;
+        private final EventExecutor executor;
+
+        private final StreamMessageAndWriter<HttpData> outputStreamWriter;
+        private final Consumer<OutputStream> outputStreamConsumer;
+        private final Executor blockingTaskExecutor;
+
+        @Nullable
+        private Subscription upstream;
+        private boolean completed;
+
+        OutputStreamSubscriber(Subscriber<? super HttpData> downstream,
+                               EventExecutor executor,
+                               StreamMessageAndWriter<HttpData> outputStreamWriter,
+                               Consumer<OutputStream> outputStreamConsumer,
+                               Executor blockingTaskExecutor) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(executor, "executor");
+            requireNonNull(outputStreamWriter, "outputStreamWriter");
+            requireNonNull(outputStreamConsumer, "outputStreamConsumer");
+            requireNonNull(blockingTaskExecutor, "blockingTaskExecutor");
+            this.downstream = downstream;
+            this.executor = executor;
+            this.outputStreamWriter = outputStreamWriter;
+            this.outputStreamConsumer = outputStreamConsumer;
+            this.blockingTaskExecutor = blockingTaskExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            downstream.onSubscribe(this);
```

Comment:
Should we just hand over `subscription` to `downstream`?
```suggestion
            downstream.onSubscribe(subscription);
```

Reply:
✅ updated!